### DISCUSSION TITLE
DateTime Range Component

### DISCUSSION THREAD

**Discussion by spatil32** [twilio-labs/paste#2125] [General] [answered]
Hi Team, I have a question about DateTimeRange component. I have a Figma design for Video Insights with a requirement to build custom DateTimeRange component. Previously in Console components, we had Advanced DateTimeRangeContainer [component](https://code.hq.twilio.com/pages/react/react-components/#/UI%20Components/Datetime%20picker) with such functionality. Is there a plan to provide such component in PASTE? Or a team should build a custom component per requirement?

![Screen Shot 2022-01-11 at 12 48 22 PM](https://user-images.githubusercontent.com/16721217/149021565-df3a4743-f323-4720-80bc-78ac7f2df8b8.png)

**Comment by nkrantz** [comment, marked as answer]
Hey @spatil32 - good question! We don't have a DateTime Picker in the system, just separate [Date](https://paste.twilio.design/components/date-picker) and [Time Pickers](https://paste.twilio.design/components/time-picker). The current guidance is to display the two components next to each other to achieve the same result as an `<input type="datetime-local />`. For ranges, that would mean 4 separate components - a start Date Picker and Time Picker, and an end Date Picker and Time Picker. On the docs pages, you can find examples of [date range](https://paste.twilio.design/components/date-picker#date-range-picker) and [time range](https://paste.twilio.design/components/time-picker#time-range-picker) implementations - you would need to combine the two for a DateTime range.

The Advanced DateTimeRangeContainer component you were previously using also has the quick-select options (`Last 6 hours`, etc.). That's another piece of functionality that was left out of scope for our current pickers, because the HTML pickers (on which our current components are built) don't support it. If it's important to have those range dropdown options, you could use a [Combobox](https://paste.twilio.design/components/combobox) with the last option as "Custom" and only display the range pickers upon selection of that option (to initially hide the clunkiness of showing all 4 pickers together). I put together a quick [CodeSandbox to demonstrate that behavior](https://codesandbox.io/s/gallant-cori-rhdom?file=/src/index.tsx).

If that option doesn't suit your use case, feel free to build a custom solution using the [Combobox Primitive](https://paste.twilio.design/primitives/combobox-primitive). Either way, let us know what direction you end up choosing!

**Comment by spatil32** [comment]
Hello @nkrantz , thank you for providing the detailed answer and sharing the code sandbox. I would build the component in a similar way with the `Custom` option opening in `Popover` with Date & Time Pickers as I have in Figma Design.

<img width="238" alt="Screen Shot 2022-01-12 at 9 45 58 AM" src="https://user-images.githubusercontent.com/16721217/149193923-eb5d2f20-2d80-46c8-9c38-432186ce47d9.png">

**Reply by spatil32** [reply]
Hello @nkrantz , I was trying to implement a custom DateTime Picker by trying to open it in Popover. However, Popover has max-width set in the component and there's no way to specify width prop by user. Because of this, I am not able to fit Date & Time picker on the same line. I saw in the discussions that similar question about Popover width limitation was asked [here](https://github.com/twilio-labs/paste/discussions/1732). Is this being worked on currently?

Here's how my component looks with default max-width:

![Screen Shot 2022-01-21 at 5 34 04 PM](https://user-images.githubusercontent.com/16721217/150619705-325045c0-6afa-44cb-a88d-0c592aaff368.png)

I tried to add some negative margin and decrease the width of Date & Time pickers so fit the selected date/time, but with the minimum possible tokens, UI still looks very crowded. If we could set width on popover, it will be easy to implement.

![Screen Shot 2022-01-21 at 5 40 23 PM](https://user-images.githubusercontent.com/16721217/150619763-c968b781-158b-41b9-9d86-36d24cb26481.png)

**Reply by nkrantz** [reply]
Hey @spatil32 - so sorry that I missed your response on this thread. I'm going to check with the team and get back to you.

**Reply by spatil32** [reply]
Hey @nkrantz , sure, please do let us know. We plan to go GA in March and design wants us to use the above mentioned DateTimePicker. I have implemented it currently as it looks in last screenshot but it's not very great looking on UI. It would be good to have some padding around and by providing prop for width for popover helps. Thank you! :)

**Reply by nkrantz** [reply]
@spatil32 that's super helpful context, thanks! I've chatted with the team and having a wider Popover is a request that we've gotten more than once recently. I [created a ticket in the backlog](https://issues.corp.twilio.com/browse/DSYS-3230) for this, feel free to follow along in JIRA. I'll try to push to get this update into our next sprint so that it's ready for you to use, but we're pretty low on capacity right now so unfortunately I can't say for sure when we'll get to it.

**Reply by nkrantz** [reply]
Hi @spatil32 - just wanted to let you know that the Popover changes are in! Keep me updated with any feedback you have 😄

**Reply by spatil32** [reply]
Hi @nkrantz , that's great! Thank you for getting this done quickly. I'll try it out today. 😃

**Reply by bhamtil-twilio** [reply]
Hi @spatil32 and @nkrantz, my team will be updating a similar component this quarter, might this be considered a pattern we can add to the library?

**Reply by nkrantz** [reply]
Hey @bhamtil-twilio - that's a good idea. We chatted as a team and agree that adding this pattern to Figma (as part of the Filter Group Pattern) would be useful. The design team is going to work on that and will update this thread once there's more info on when you can expect to see it!

**Reply by kate-porter** [reply]
Hey @bhamtil-twilio! Just wanted to let you know we've just published this custom DateTime Popover in the Paste Patterns library within the Filter group pattern. Let us know if you have any questions or feedback!
